File: two_sum/two_sum.pl
twoSum([],Target,[]).
twoSum([H|T],Target,X) :- tsHelper(T,H,Target,X).

tsHelper([],_,_,[]).
tsHelper([X],Y,T,[Y,X]) :- T=X+Y.
tsHelper([_],_,_,[]).
tsHelper([H|T],X,Target,[X,H]) :- Target==H+X.
tsHelper([H|T],X,Target,S) :- tsHelper(T,X,Target,S), S=[_|_].
tsHelper([H|T],X,Target,S) :- tsHelper(T,H,Targt,S).


% funktioniert nicht

File: two_sum/two_sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    hash = {}
    nums.each_index do |i|
        x = target - nums[i]
        return [hash[x], i] if hash.has_key?(x)
        hash[nums[i]]= i
    end
    return nil
end


# Faster Solution:
# # @param {Integer[]} nums
# # @param {Integer} target
# # @return {Integer[]}
# def two_sum(nums, target)
#     complement_index = {}
#     nums.each_with_index do |num, index|
#         if complement_index[num]
#             return [index, complement_index[num]]
#         else
#             complement_index[target - num] = index 
#         end
#     end
# end

